activerecord/lib/active_record/timestamp.rb

```diff
@@ -83,6 +83,17 @@ def timestamp_attributes_for_update
         def current_time_from_proper_timezone
           default_timezone == :utc ? Time.now.utc : Time.now
         end
+
+        def _insert_attributes(attributes)
+          super
+
+          attributes.map do |insert_attribute|
+            current_time = current_time_from_proper_timezone
+            timestamp_attributes = all_timestamp_attributes_in_model.index_with(current_time)
+
+            timestamp_attributes.merge(insert_attribute.with_indifferent_access)
```

Comment:
@kaspth I had to use `insert_attribute.with_indifferent_access` here so that we can override timestamp attributes if they are given already. Is this the correct way?

Reply:
This isn't right. Thinking about this some more we shouldn't use `merge` at all, we should add an extra lookup in insert_all: `bind = Relation::QueryAttribute.new(key, attributes[key] || defaults[key], types[key])`

And then we'd revise the solution to use:

```ruby
# In persistence.rb
def insert_all(…)
  InsertAll.new(…, defaults: _insert_defaults).execute
end

def _insert_defaults
  {}
end

# In timestamp.rb
def _insert_defaults
  if record_timestamps
    super.merge(self.class.touch_attributes_with_time(timestamp_attributes_for_create_in_model))
  else
    super
  end
end
```

Then we'd have a corresponding `_upsert_defaults`.